Mathlib/Topology/CWComplex/Classical/Basic.lean

```diff
@@ -523,38 +523,220 @@ lemma CWComplex.cellFrontier_subset_finite_openCell [CWComplex C] (n : ℕ) (i :
+      rw [inter_eq_right.2 this]
+      exact isClosedBase C
+    intro n _ j
+    by_cases h : j ∈ I n
+    · right
+      suffices closedCell n j ⊆ E by
+        rw [inter_eq_right.2 this]
+        exact isClosed_closedCell
+      exact closedCell_subset n ⟨j, h⟩
+    · left
+      rw [← union, union_inter_distrib_right]
+      apply IsClosed.union
+      · rw [inter_comm, (disjointBase n j).inter_eq]
+        exact isClosed_empty
+      · simp_rw [iUnion_inter]
+        suffices ⋃ m, ⋃ (i : ↑(I m)), openCell m (i : cell C m) ∩ openCell (C := C) n j = ∅ by
+          rw [this]
+          exact isClosed_empty
+        apply iUnion_eq_empty.2 fun m ↦ iUnion_eq_empty.2 fun i ↦ ?_
+        exact (disjoint_openCell_of_ne (by aesop)).inter_eq
+  union' := union
+
+/-- An alternative version of `Subcomplex`: Instead of requiring that `E` is closed it requires
+  that for every cell of the subcomplex the corresponding closed cell is a subset of `E`. -/
+@[simps! -isSimp]
+def CWComplex.Subcomplex.mk' [T2Space X] (C : Set X) [CWComplex C] (E : Set X)
+    (I : Π n, Set (cell C n))
+    (closedCell_subset : ∀ (n : ℕ) (i : I n), closedCell (C := C) (D := ∅) n i ⊆ E)
+    (union : ⋃ (n : ℕ) (j : I n), openCell (C := C) (D := ∅) n j = E) : Subcomplex C :=
+  RelCWComplex.Subcomplex.mk' C E I closedCell_subset (by rw [empty_union]; exact union)
+
+/-- An alternative version of `Subcomplex`: Instead of requiring that `E` is closed it requires that
+  `E` is a CW-complex. -/
+@[simps -isSimp]
+def RelCWComplex.Subcomplex.mk'' [T2Space X] (C : Set X) {D : Set X} [RelCWComplex C D] (E : Set X)
+    (I : Π n, Set (cell C n)) [RelCWComplex E D]
+    (union : D ∪ ⋃ (n : ℕ) (j : I n), openCell (C := C) (D := D) n j = E) : Subcomplex C where
+  carrier := E
+  I := I
+  closed' := isClosed (D := D)
+  union' := union
+
+/-- An alternative version of `Subcomplex`: Instead of requiring that `E` is closed it requires that
+  `E` is a CW-complex. -/
+@[simps -isSimp]
+def CWComplex.Subcomplex.mk'' [T2Space X] (C : Set X) [h : CWComplex C] (E : Set X)
+    (I : Π n, Set (cell C n)) [CWComplex E]
+    (union : ⋃ (n : ℕ) (j : I n), openCell (C := C) (D := ∅) n j = E) :
+    Subcomplex C where
+  carrier := E
+  I := I
+  closed' := RelCWComplex.isClosed (D := ∅)
+  union' := by
+    rw [empty_union]
+    exact union
+
+end Subcomplex
+
+section skeleton
+
+variable [T2Space X]
+
+namespace RelCWComplex
+
 /-- A non-standard definition of the `n`-skeleton of a CW complex for `n ∈ ℕ ∪ {∞}`.
 This allows the base case of induction to be about the base instead of being about the union of
 the base and some points.
 The standard `skeleton` is defined in terms of `skeletonLT`. `skeletonLT` is preferred
 in statements. You should then derive the statement about `skeleton`. -/
-def RelCWComplex.skeletonLT (C : Set X) {D : Set X} [RelCWComplex C D] (n : ℕ∞) : Set X :=
-  D ∪ ⋃ (m : ℕ) (_ : m < n) (j : cell C m), closedCell m j
+@[simps! -isSimp, irreducible]
```

Comment:
If that's not too hard to do, could you set up the `simps` attribute so that the `carrier` spelling is replaced by the coercion coming from `SetLike`?

Reply:
What I meant is that we want the canonical spelling for "the set underlying a subcomplex `E`" to be `↑E` rather than `E.carrier`, hence we want the lemmas generated by `@[simps]` to use that spelling.

I did not realize this was already the case just thanks to the `SetLike` instance (this is also dark magic to me, I didn't expect it to be this sophisticated!).

I guess the question of the lemma names remain: to be concrete, I think I would prefer the lemma `skeletonLT_carrier` to be called `coe_skeletonLT`, since the lemma is not about `(skeletonLT C n).carrier`. If you agree, you should add the following right after the `SetLike` instance:
```lean
initialize_simps_projections Subcomplex (carrier → coe, as_prefix coe)
```